### DISCUSSION TITLE
Return the old "edit" button feature in the main menu

### DISCUSSION THREAD

**Discussion by Mechanizen** [ppy/osu#27990] [Ideas]
Something that really throws me off in lazer is that to edit an already existing beatmap, you have to do **Play>Solo>Select a beatmap>Right Click>Edit**. Here you can visually see how tedious it is compared to stable where you just do **Edit>Double click on beatmap** and you are already in the editor.

That's really not smooth at all and it could even induce people into thinking that you cannot edit already existing beatmaps.
Like when i click on **Edit**, there is no option to edit an already existing beatmap because it is only available in **Play>Solo**. There is something sementically wrong there, if there is an **Edit** tab in the main then why do I have to go through **Play>Solo** to access that editor feature?

Also even if we got used to go through **Play>Solo**, we would still have to do the **Right Click>Edit** thing which in my opinion is quite a downgrade in terms of user experience compared to just double clicking.

So the request here is to return the "old edit button" on the main menu. Maybe add a tab in the **Edit** menu labeled "Edit a beatmap" and rename the already existing tab (currently named "beatmap editor") to "create a new beatmap" (which is essentially what it is).

For personnal context, as a mapper, not being able to just go from the main menu to **Edit** then just browse beatmaps and open some of them in the editor with double click makes me feel homeless. In stable you clearly have "play" and "edit", as if you had the "visitor entrance" and the "artist entrance" in a club. As a mapper, not having my "entrance" in Lazer feels very off.

**Comment by peppy** [comment]
We intend to implement a similar editor selection interface but I didn't want to use the main song select view for this because it feels weird.

**Reply by murryguy** [reply]
O.K.?

**Comment by joppyloppy** [comment]
I would also appreciate the 'Edit' button on the main menu sending me to a 'list of editable beatmaps', rather than sending me directly to the 'setup' screen for a new, non-existing beatmap.

**Comment by peppy** [comment]
@vatei this might be an area you can help with design proposal to get things moving.

a screen when arriving to the editor from the main menu that lets you see recently edited beatmaps, maps created by you and a button to start fresh.

**Reply by vatei** [reply]
Here's what I got so far 👍 

![image](https://github.com/user-attachments/assets/6623226f-3bcb-4ce7-9463-049af3950eba)
![image](https://github.com/user-attachments/assets/f16bdea1-de64-49c2-95d4-ac2347246bfc)

https://github.com/user-attachments/assets/256de463-7670-4002-acba-2bfbbd980f8a


The "new project" button brings to the editor setup screen
The "open existing" button brings to the carousel, to select a beatmap
The search bar only searches through edited maps
Users can organize their edited maps into folders, and can import maps from a collection they have in song select

---
Originally I had this, but I ran into a few issues with this iteration, especially after getting some feedback from a few players

![image](https://github.com/user-attachments/assets/d8c02612-e2f8-4d34-bf4a-a561d3b98206)

- Use feels limited, especially considering mappers' existing workflows
- Original beatmap card designs from web don't work well for this. The purpose is to go to the editor screen, yet the edit button is tucked on the side and is very small. 
- I wanted to be able to click directly on the card to go to the editor, but original card has a lot of clickable elements, which make it hard to know where to click
- Overall felt barebones

As a result, this proposal includes a feature to let users organize their edited maps into folders, which feels like an useful and convenient addition to this screen.
This is also styled with v2 designs in mind.

I hope that's acceptable, otherwise I'll iterate accordingly 👍

**Reply by tadatomix** [reply]
Maybe the Activity feed should've been put somewhere else? I don't think it's actually the first thing people might want to see, but I can be wrong

**Reply by Mechanizen** [reply]
Maybe it's because of these particlar screenshots which have a lot of notifications. It all comes down to : "what is going to be displayed on the Feed?". If only major notifications like Nominations / Qualification, etc... Are displayed it will not get as cluttered as these screenshots. However if we starting including notifications from modding, comments, likes and suchlikes it's going to be a problem very fast.

_So I would agree that the feed part is cluttering the UI / is not the primary thing that a mapper is looking for._

**Open Existing feature**

I could add that the "open existing" button which is basically the old Edit button in the Main Menu should be at the bottom right. People (including myself) might not get used to this new "intermediate" screen easily. Placing the "Open Existing" button at the bottom right with a pretty large size would help redevelop muscle memory more smoothly and prevent us from having to search our way in the UI everytime.

Also imagine that if someone is searching for a specific map in folder but cannot find it, the use will want to browse all the existing beatmaps to find it. If the "Open Existing" button is at the top, the user will have to scroll back up (and do some menuing if they in a submenu). Having this button at the bottom right at all times would provide an "escape button" when we get lost and reduce the frustration.

**New Project feature**

Moreover, if we move the "Open Existing" button to the bottom right, the "New Project" button will be alone and cold near the search bar while it is an important feature. I would reckon making it bigger, spanning the whole width of the scroll panel and move it above the search bar.

**Search Bar**

As a result from the previous remark, extend the search bar to fill the space left by displacing the two buttons. Possibly make it a bit taller so that it's less frustrating to click on and easier to spot.

**For future developments**

If the previous comment is applied, the search bar may not need the whole length of the scroll panel. So now you have the opportunity to add other features to the left of the search bar to improve the "search experience" (like filters and whatnot).

**Reply by Mechanizen** [reply]
Here is a pretty bad paintjob of something that looks better to me:

- The **New Project** button has been shifted to the top and allocated more space
- The **Open Existing** button has been shifted to the bottom right for more convenience
- The search bar has been slightly extended in length and height
- New search features can be added to the right of the search bar (tags, filters, ...)
- The **Recently Edited** section has been moved above the feed

The last thing that is puzzling me are the two "Home" and "All beatmaps" tabs that look kinda lost in all of this.
Could it be possible to consider a sidebar for this kind of navigation tabs?
(Or at least find a way to give them more weight within the UI because we barely notice them right now)

![newui](https://github.com/user-attachments/assets/b1ee87fe-38c9-4f5e-912d-ce3492b78c40)

**Reply by vatei** [reply]
Thanks for the notes, I'll look at them in detail as soon as I go back to this!
I just want to wait for peppy's green light as some of the proposed features might not fit his vision or be out of scope for this issue.
Especially the folder feature, feels a bit like having collections but different somehow. I wonder if there's a way to leverage the existing collection system in an intuitive way on this screen?
Edit: I somehow never used or noticed the compact view on the beatmap browser, that redesign might not be needed 😓
Sorry about that, I'll try using it instead (I feel very stupid)

**Reply by vatei** [reply]
Did not address most of the previous concerns (yet), just fixed basic mistakes

![image](https://github.com/user-attachments/assets/1e79163e-4640-4784-a24b-be942108a3c1)

**Reply by bdach** [reply]
a few problems I have with these

- all of the designs show sets not individual difficulties as the points of entry into editor proper. that doesn't work for me progression wise - how do you edit a particular difficulty of a beatmap set like this? it's especially important given that a significant percentage of mappers do guest sets now
- what does the "open existing" button even do
- "new project" doesn't seem appropriately exposed
- search box position is weird. what is the search box even searching? the entire screen?
- maybe not immediately but I would like this screen to also contain a way to use featured artist map templates so that someone can start mapping them instantly
- I have no idea what the folder thing is for
- I have doubts about the usability of the activity feed as presented. I dunno whether mappers need to know whether their map got moved to qualified or whatever that much. what they will more likely care about is whether they had any mods posted on them. such maps should be prioritised display wise because they're the ones that the user is most likely to want to edit

**Reply by vatei** [reply]
I have to agree, I would have reused the dropdown to select the wanted difficulty, and it would only show local ones, but maybe a different design is needed
I'll try to think of something better 👍 

> * "new project" doesn't seem appropriately exposed

Could you give an example? 

> * search box position is weird. what is the search box even searching? the entire screen?

All local maps

> * I have no idea what the folder thing is for

To organize maps instead of having a big list. Although it feels like a worse duplicate of a collection.
I was thinking about just reusing collections instead but presenting them in this folder form?
Meaning, clicking the + button from this screen would create a collection, which sounds better since you could also see these groups in song select.
Does that sound better?

I'll rethink everything, thanks for the feedback. The featured artist idea sounds good.

**Reply by bdach** [reply]
> > * "new project" doesn't seem appropriately exposed
> 
> Could you give an example?

it just should be much more visible. right now sure it's the biggest button but it's shoved in a corner and not really that exposed. the other thing is that you should be able to just drag and drop an .mp3 which also needs to be spelled out somewhere

> > * search box position is weird. what is the search box even searching? the entire screen?
> 
> All local maps

with the positioning of it it's weird, I'd honestly expect it to encompass the entire screen, e.g. also search the activity feed and folders or whatever

> To organize maps instead of having a big list. Although it feels like a worse duplicate of a collection. I was thinking about just reusing collections instead but presenting them in this folder form? Meaning, clicking the + button from this screen would create a collection, which sounds better since you could also see these groups in song select. Does that sound better?

I still think the value of this is dubious because I haven't seen anyone seriously ask for something of this kind but I'd much rather repurpose collections for this yes

**Reply by ziv-vy** [reply]
There should also probably be added a stable-like featured artists button

**Reply by tadatomix** [reply]
> There should also probably be added a stable-like featured artists button

It shouldn't be a button, but a tab/section

**Reply by vatei** [reply]
I changed a bunch of things
- Added a ruleset selector, thought it might be useful to have in the landing screen
- The "home" tab now contains an area to drop a .mp3, and a big button to use a template. For now, this button would redirect to the featured artist listing, but could later on redirect to a "featured artist" tab inside the landing page.
- Added two sections, one for recently edited maps and one for last modding updates.
- The beatmap cards are now split into individual difficulties.
- Added notification indicators for problems, notes, suggestions, and praise. Right now, clicking them wound redirect to the beatmap discussion page, but a separate "modding" tab could be nice to have in the future?
- The "all beatmaps" tab now contain all maps edited by the user, grouped by collections, with a default one for no collections.
- Search would still show the results grouped by collections, same for sorting options.
- Manage collections opens the manage collection window, like in song select, but only displays collections created in this screen.
- Collections created in this landing page would still show up in song select
- You can collapse the collection sections

https://github.com/user-attachments/assets/ddbae123-e0c4-4442-962e-2d301b647cea

![image](https://github.com/user-attachments/assets/f86e1bbc-507c-49fc-91bd-fe9aa7b2fba9)

![image](https://github.com/user-attachments/assets/2d3ba35a-4f4b-4e4c-b90d-7d19eeadbe08)

And with the potential other tabs:

![image](https://github.com/user-attachments/assets/4dd4d1e1-2ed2-47b9-b468-a7e44946ccc7)


I'm hoping this is more in line with mapper's expectations 👍 
I'll admit I'm not super in touch with the mapping side of the game so this might take a few more iterations?
Eager to hear some feedback!

Edit: as an alternative this screen could just use song select like suggested above

![image](https://github.com/user-attachments/assets/3afbc61f-4a2e-4120-b86e-44dcecc03fac)

**Reply by tadatomix** [reply]
Looking better, but what is the difference between "Recently added" and "Last update"? And, to be honest, I'm not sure why we are still stuck with osu-web cards for difficulties

**Reply by Mechanizen** [reply]
_Long one sorry_

I like the **Browse Song Select** at the bottom.

**Game mode tabs**
The game mode selection tabs at the top look out of place to me. What is difference in UI when we select different game modes? If the point is to select recently edited beatmaps for each individual mod then it's going to be very annoying for anyone not mapping std. They will have to select back their main game mode everytime.

Imo you should have only one screen where all difficulties of all game modes are displayed at the same time. You could have some filters but they should be optional, you shouldn't be forced to display a particular game mode in priority.

**Arrows**
Also only displaying 3 recently edited beatmaps per category feels very underwhelming to me. I don't like the idea of having an arrow on the right to show more maps. I'm certain most people will not use it and just skip to "Browse song select" right away and type the name of the map they want. Also the arrows have that grey fade to their left. Having an unclear UI boundary between the arrow and the underlying beatmap thumbnail is a huge missclick trap. People are going to click the beatmap instead of the arrow 50% of the time.

I would much rather having two rows of "recently edited" difficulties and having to scroll a tiny bit to find what I want than clicking that arrow.

**Last Updated Maps**
Remove the "last updated difficulties" section which feels like a duplicate of the previous "recently edited beatmap". If you update a map it implies that you have modified it, nobody would update a map without modifications. Both sections are going to display exactly the same maps.

**"Your beatmaps"**
At this point, could it be possible to find all beatmaps (stored locally) that were created by the currently logged-in player?

I think it would make much more sense to have just at the top the "recently edited difficulties" displaying the last 6-8 difficulties that were edited and below a list of all maps created by the player under a "your projects" section.

In the "Your projects" section you would have thumbnails designating mapsets. If you click on one of the mapset thumbnail, an overlay is displayed to select the target difficulty to edit. If the mapset contains only a single difficulty then open that difficulty directly.

I would love to see a "your beatmaps" feature as it's currently missing from stable whilst being an essential feature. Currently in stable you need to either:

- Search your name in the song select (but if your name appears elsewhere, irrelevant maps will show up so you need to use "mapper=name" which is super annoying)
- Change the sorted filter to group maps by author and then search for your name

And even then, you may not be able to find what you want because if your map is a GD, it may not show up.

Having a dedicated feature that displays all projects created by the current user, including GDs in other mapsets would be huge to keep track of ongoing projects.

I don't think that thinking this menu strictly in terms of a "Quick Access" menu will bring a lot of adhesion from the community. People are already used to find their maps in the song-select screen so adding an intermediate screen will not add a lot of value to the process if not introducing more clicks. 

**Turning this screen into a "Project Control Panel" where the player can have a clear view of all the maps belonging to them would be of great value as every maps are currently scrambled together in stable.**

**Remark:** I think this screen should only display maps that BELONG to the mapper (sets, difficulties and GDs). If the mapper opens someone else's map in the editor, it should not show up on that screen.

**Reply by tadatomix** [reply]
@vatei, sorry for ping, but that's how I might want to see this startup screen. It's **really raw** and made it by remixing your and flyte's parts, but since I consider it as a reference, I think, it's quite okay
![beatmap editor](https://github.com/user-attachments/assets/6207069c-b140-4a63-96ae-1ccb8883f720)
To you and anybody else, feel free to share your opinions, I just gave a suggestion of a design

Why have I did it that way? I still think cards are bad for this screen, when the person wants to see difficulties separetely. I really don't know where and how to put betmap modding stuff, so I did my best here. Also, I don't have Torus Pro, so I'm using it's demo version. It lacks all of the punctuation glyphs, so that's why you don't see me using majority of them. And, personally, I think, having or between these 2 places is more common and okay, so that's why I did it that way

UPD: Mechanizen is right about the last updated difficulties section. How about replacing it with "Recent mods"? Like have the same card idea, but like "X said under Y map Z diff "You need to change α"

**Reply by vatei** [reply]
I'll read the rest later but the selector was more meant to know the ruleset you're in (and change it) when you create a new map. 
When you create a new map from the edit button, it already uses the ruleset you were in before, but it doesn't show it explicitly until you get to the editor, which I found quite confusing
I wasn't planning to only limit rulesets displayed on screen to one if that makes sense.

**Reply by Mechanizen** [reply]
@tadatomix This looks more it yes, I love it

@vatei That makes sense, if it's that way it's good although I still think that displaying everything on the same page and distinguishing rulesets by symbol as tadatomix proposed is a better way of doing it. People rarely switch rulesets in mapping, I don't see the benefit if segragating rulesets on that menu.

**Reply by tadatomix** [reply]
@Mechanizen thanks for your opinion. Tried my best at giving it a better feeling, for me.

**Reply by vatei** [reply]
Okay so a few points @tadatomix @Mechanizen 
- Did you guys take a look at the "all beatmaps" tab? Might address the idea about having a place to see all edited beatmaps, they are even organized by collection.
- Cards are already split into single diff here. You can see in my screenshots that there is only one difficulty in the card
- Maybe there is another way to do it, but right now, if you drag and drop a song or use a template, it will always use the previously selected ruleset as the default beatmap. Like I said before, I had no intent to segregate diffs from other rulesets away, you can even see I added the ruleset icon inside the beatmap card. I should have put more examples but yeah.
- Last updated is meant to be last discussed/modded. Bad wording on my part, sorry about that, open to suggestions 
- I like the proposed design, feels closer to the editor, I wanted to go more in line with v2 as the game is being updated with v2 designs. I'll try using the song select designs for the cards, but as you saw that forces you to have a vertical layout. I also updated by accident the cards designs (then went back to osu web's), maybe you would like them more?
![image](https://github.com/user-attachments/assets/1ac1d7f9-1e2c-4c96-bc25-f12c7ff61cda)

**Reply by tadatomix** [reply]
![beatmap editor_new](https://github.com/user-attachments/assets/16de1318-39f6-4370-bb85-ffdbe0311d5a)
I've made my concept with recent posts, by the way

**Reply by tadatomix** [reply]
@vatei 

- "All maps" is quite better, yet so, my concern with cards is lack of the free space, if you get what do I mean. As you can see, on my screenshot I can have 8 maps, when you home screen fits only 3 maps. The problem with this is that you need to spam many times LMB to edit a map, you've had week ago, if you opened 3+ maps after that one. At this moment, even going to song select and RMBing at the diff of your preference, to edit, is less convoluted.
- Splitting cards by a single diff is trying to kill 2 birds with a single stone, yet you have -5 dioptries. I'm sorry, but that just making both things quite worse. Card's purpose was to group all of the maps inside the mapset in a single place. Making a single difficulty as a card is strange, for me
- This one, seems like, was addressed to Mechanizen, so I can't give a strong opinion here
- I gave a suggestion, to have a listing of recent mods, like I made on the screenshot on the post above
- Thank you for giving your opinion on my design. These cards looking a bit better, my only suggestion is invert the colors, so I think make the background in color of the star rate is better than having a small colored star rate pill

And a small mod for your design. As long as .wav really supported by lazer as the audio format, it's a **raw audiofile** that means that an average 3 minute song will be 50 megabytes. I don't think you should encourage people to use wavs for these cases, since .ogg pretty does it's work at giving the best possible audio quality for it's lossy 192kbps

**Reply by vatei** [reply]
I'll try something with the vertical layout.
Noted for .wav
Should I reuse the greenish design or keep going with what I have?

**Reply by tadatomix** [reply]
>Should I reuse the greenish design or keep going with what I have?

If you want to. I used osu!editor setup screen palette, since I think it's more logical, in case with a screen, exclusively for osu!editor

**Comment by Mechanizen** [comment]
Since @vatei is exploring this issue with the idea of a "dashboard interface", I will write here some thoughts about it in the light of my initial post. This isn't really a criticism of the ongoing development, more like a general reflection on the direction we are taking.

My initial post stated: 

`Something that really throws me off in lazer is that to edit an already existing beatmap, you have to do **Play>Solo>Select a beatmap>Right Click>Edit**. Here you can visually see how tedious it is compared to stable where you just do **Edit>Double click on beatmap** and you are already in the editor.`

My concern was about the accessibility of the editor in lazer and the fact that not having the old "Edit Beatmap" button was forcing the mapper to do more clicks to edit a beatmap.

The idea of implementing a "dashboard interface" when clicking on **Editor** in the main menu can be a good idea but there are high risks that it does not address the initial issue which was fast accessibility to the editor. _In fact, adding an intermediate menu between the song selection carroussel and the main menu is only adding more clicks from the main menu to the editor._

To provide fast access to the editor, the idea of adding "recently edited beatmaps" sections to the dashboard is good. But in order to provide at least a similar experience as the legacy way of navigating to the editor, the "recently edited beatmaps" feature should:

1. Display enough maps to allow the mapper to reach far enough in their editor history
2. Be compact enough to ease the visual search of items
3. Necessitate minimal menuing to find a beatmap (no additionnal clicks to visually find an item, no extensive scrolling)

Having a proper fast access feature is essential but I don't think it will be enough to gain adhesion from the community. The current way of accessing the editor in lazer **Play>Solo>Select a beatmap>Right Click>Edit** is longer than the stable one but it's not too bad. People could get used to it even though it takes a bit more time to get to the editor.

If our dashboard is only about providing an history of recently edited beatmaps, I do not think it will be enough of a justification for mappers to start using it. This dashboard has to bring more value that just a seemingly "quick access". By clicking **Editor** in the main menu, the mapper should be able to access features that are not already easily accessible when clicking on the **Song Selection** button in the main menu.

Let alone, there are risks that the mapper cannot find what they are searching for on the dashboard. As the "recently edited beatmaps" feature only displays a small set of beatmaps, it is possible that what the mapper is searching for will not be displayed. In this case they will have to navigate back to the main menu and into song selection to find what they want, inducing frustration. If this happens too often, the users will simply stop using the dashboard and go to song select everytime.

This is why I proposed to add  the "Go to song selection" button at the bottom right. This way the user has a panic escape button when they cannot find what they want, minimizing the frustration.

To add value to the dashboard, we should implement features that are missing from stable:

### Display all maps created by the player

Currently in stable, all maps are mixed together in the carroussel and it is difficult to keep track of all the maps that were created by the mapper. To find them you have multiple options like searching for your name or sorting beatmaps by mapper. Although this way of researching for your own maps is not satisfactory: 

- if your name or part of your name appears in the title of a song or in the title of a difficulty, irrelevant results will be displayed
- if the map you are searching for in a GD and the difficulty name does not contain your username, it will not be found

Currently, in stable or lazer, there is no satisfying way of displaying all "ongoing projects" as everything is mixed together with other maps in the song selection carroussel. 

**The spirit of the editor dashboard should be to provide a clear view of all medias (maps, sets, GDs, audio files, ...) belonging to the mapper.**

### Provide space for future editor features

Alongside finding a beatmap, the dashboard should also make space for additionnal features related to mapping. This way, the dashboard is not just an alternative version of the song selection menu.

Some features going in that direction have already been proposed and implemented:

- News feed to track activity on the mapper's submitted beatmaps
- Access to featured artist audios and templates

I haven't kept up with Lazer in a long time but I think there were also plans to introduce features like "saving patterns" / "saving map snippets". The dashboard should also have as a duty to host these in a library tab.

The dashboard could also provide a visual interface to see manage assets in a library fashion : audio effects for hitsounding, visual effects and sprites for storyboarding, backgrounds, unmapped mp3s, ...

I saw that there is also development in editor customization with custom snapping grids. The dashboard should also allow the mapper to manage its setting, make and manage custom grids. 

The dashboard could also allow to make and manage difficulty presets for the user to quickly setup new beatmaps:

- Name the presets
- Create difficulty presets (HP, AR, CS, ...)
- Colors presets
- Grid presets
- Hitsound presets

### In short

**In essence, in order for the dashboard to be useful and accepted, it should provide more than just a quick access to beatmaps feature.**

The fast access to "recently edited beatmaps should":

1. Display enough maps to allow the mapper to reach far enough in their editor history
2. Be compact enough to ease the visual search of items
3. Necessitate minimal menuing to find a beatmap (no additionnal clicks to visually find an item, no extensive scrolling)
4. Display only the player's unranked maps

**The dashboard should provide features that are not accessible / not easily accessible in the song selection menu related to mapping specifically:**

1. Notification feed for submitted maps
2. Easy access to featured artist templates and mp3
3. Other features related to asset management
4. Leftover space for future developments

**The dashboard should eventually turn into an "editor homepage" and not just be a song selection V2.**
